src/transports/http.rs

```diff
@@ -1,283 +1,173 @@
     }
 
-    fn send_request<F, O>(&self, id: RequestId, request: rpc::Request, extract: F) -> Response<F>
-    where
-        F: Fn(Vec<u8>) -> O,
-    {
-        let request = helpers::to_string(&request);
-        log::debug!("[{}] Sending: {} to {}", id, request, self.url);
-        let len = request.len();
-
-        let mut request_builder = self
-            .client
-            .post(self.url.clone())
-            .header(
-                reqwest::header::CONTENT_TYPE,
-                HeaderValue::from_static("application/json"),
-            )
-            .header(reqwest::header::USER_AGENT, HeaderValue::from_static("web3.rs"))
-            .body(request);
-
-        // Don't send chunked request
-        if len < MAX_SINGLE_CHUNK {
-            request_builder = request_builder.header(reqwest::header::CONTENT_LENGTH, len.to_string());
+    /// Like `new` but with a user provided client instance.
+    pub fn with_client(client: Client, url: Url) -> Self {
+        Self {
+            client,
+            inner: Arc::new(Inner {
+                url,
+                id: AtomicUsize::new(0),
+            }),
         }
+    }
 
-        // Send basic auth header
-        if let Some(ref basic_auth) = self.basic_auth {
-            request_builder = request_builder.header(reqwest::header::AUTHORIZATION, basic_auth.clone());
-        }
+    fn next_id(&self) -> RequestId {
+        self.inner.id.fetch_add(1, Ordering::AcqRel)
+    }
 
-        let result = request_builder.send();
-        Response::new(id, Box::pin(result), extract)
+    fn new_request(&self) -> (Client, Url) {
+        (self.client.clone(), self.inner.url.clone())
     }
 }
 
+// Id is only used for logging.
+async fn execute_rpc<T: DeserializeOwned>(client: &Client, url: Url, request: &Request, id: RequestId) -> Result<T> {
+    log::debug!("[id:{}] sending request: {:?}", id, serde_json::to_string(&request)?);
+    let response = client
+        .post(url)
+        .json(request)
+        .send()
+        .await
+        .map_err(|err| Error::Transport(format!("failed to send request: {}", err)))?;
+    let status = response.status();
+    let response = response
+        .bytes()
+        .await
+        .map_err(|err| Error::Transport(format!("failed to read response bytes: {}", err)))?;
+    log::debug!(
+        "[id:{}] received response: {:?}",
+        id,
+        String::from_utf8_lossy(&response).as_ref()
+    );
+    if !status.is_success() {
+        return Err(Error::Transport(format!(
+            "response status code is not success: {}",
+            status
+        )));
+    }
+    helpers::arbitrary_precision_deserialize_workaround(&response)
+        .map_err(|err| Error::Transport(format!("failed to deserialize response: {}", err)))
+}
+
+type RpcResult = Result<Value>;
+
 impl Transport for Http {
-    type Out = Response<fn(Vec<u8>) -> error::Result<rpc::Value>>;
+    type Out = BoxFuture<'static, Result<Value>>;
```

Comment:
The goal of `type Out` on `Transport` was to avoid boxing, is it still possible to avoid?

Reply:
The previous code was not avoiding boxing. Probably an earlier version with hyper was but not currently. While `Response` did not have to be boxed, its field `state: ResponseState` was an enum of two boxes. So this code reduces the number of boxes by 1.
It probably would be possible to avoid boxing with a custom Future implementation again but considering we are sending network data, serializing data, storing a response in a Vector etc I do not think it is worth the complexity. The nice way to avoid boxing here in the future is with `type Out = impl Future<Item=Result<Value>>` but impl trait on associated types is unstable.
Anyway I would not like to do that as part of this PR as the code was already boxing but I will consider it for another one.

Reply:
Fair enough, sounds good, the extra PR is optional if we were already `Box`ing.